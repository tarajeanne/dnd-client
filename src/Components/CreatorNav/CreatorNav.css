* {
  box-sizing: border-box;
}

a {
  color: white;
}

a:hover {
  color: #ffd966;
}

.character-creator-nav {
  width: 100%;
  display: flex;
  flex-flow: row nowrap;
  background-color: #403e41;
  border-radius: 0 0 90px 0;
  justify-content: flex-end;
}

.creator-nav-content {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
  align-items: center;
}

nav {
  font-family: 'Montserrat', sans-serif;
}

.active {
  font-weight: 800;
  background-color: #2d2c2e;
}

.spacer {
  display: inline-block;
  padding-bottom: 10px;
  width: 100%;
  height: 100%;
  background: linear-gradient(to right, #ffbe00, #e85b2c);
  box-shadow: 0 2px 2px 2px gray;
  margin-bottom: 30px;
  border-radius: 0 0 90px 0;
}

.fa-edit {
  font-size: 16px;
  margin-left: 3px;
}

.fa-edit:hover {
  color: #ffd966;
}

.name-edit {
  font-size: 32px;
}

.alert {
  color: #e07450 !important;
  font-weight: bold;
  margin-left: 1px;
}
.character-creator-header {
  color: white;
  font-size: 32px;
  padding: 30px 0;
  display: inline-block;
}

.nav-area {
  display: inline-block;
  width: 100%;
  margin-bottom: 10px;
}

.hamburger {
  height: 35px;
  margin-left: 5px;
}

.nav-menu {
  margin: 15px 0 0;
  text-align: left;
  position: absolute;
  left: 30%;
  background-color: #403e41;
  opacity: 0.98;
  padding: 10px;
  box-shadow: 0 0 4px 2px gray;
  z-index: 100;
  border-radius: 5px;
}

.nav-menu > li {
  list-style: none;
  margin: 5px 2px;
  padding: 10px;
  font-size: 16px;
  border-radius: 5px;
}

.nav-menu > a {
  text-decoration: none;
}

.nextback {
  text-align: center;
  margin: auto;
  position: relative;
}

.hidden {
  display: none;
}

button {
  vertical-align: middle;
  font-size: 18px;
  padding: 5px 15px;
  border-radius: 10px;
  background: linear-gradient(to left, #ffbe00, #e85b2c);
  color: white;
  border: none;
}

.close-nav {
  position: absolute;
  top: 7px;
  right: 7px;
  color: white;
}

button:hover {
  cursor: pointer;
}

.nextback-button {
  width: 80px;
  height: 35px;
  margin-left: 5px;
  margin-right: 5px;
}

@media print {
  .Header *,
  .creator-header *,
  footer {
    display: none !important;
  }
}

@media screen and (min-width: 500px) {
  .spacer {
    width: 50%;
  }
  .creator-nav-content {
    width: 100%;
    max-width: 550px;
  }
}
