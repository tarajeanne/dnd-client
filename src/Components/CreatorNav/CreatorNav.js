import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import CharacterContext from '../../contexts/CharacterContext';
import CharacterApiService from '../../services/character-api-service';
import './CreatorNav.css';

class CreatorNav extends React.Component {
  constructor(props) {
    super(props);
    this.node = React.createRef();
    this.menu = React.createRef();
    this.state = {
      screen: this.props.match.path.split('/')[3],
      characterId: this.props.match.path.split('/')[2],
      baseUrl: this.props.match.path
        .split('/')
        .slice(1, 3)
        .join('/'),
      editingName: false,
      showNav: false
    };
  }

  componentWillMount() {
    document.addEventListener('mousedown', this.handleMouseDown, false);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleMouseDown, false);
  }

  static propTypes = {
    match: PropTypes.shape({
      path: PropTypes.string
    })
  };

  handleMouseDown = (e) => {
    e.stopPropagation();
    if (this.node.current) {
      console.log(this.node.current.contains(e.target));
      if ((this.node.current.contains(e.target) ||
      this.menu.current.contains(e.target))) {
        console.log('this part ran');
      }
      else if (this.state.showNav) {
        this.toggleNav();
      }
    } 
  };

  handleNameClick = () => {
    this.setState({
      editingName: !this.state.editingName,
      name: this.context.character.name
    });
  };

  toggleNav = () => {
    this.setState({
      showNav: !this.state.showNav
    });
  };

  handleNameChange = (e) => {
    this.setState({
      name: e.target.value
    });
  };

  handleNameSubmit = (e) => {
    e.preventDefault();
    const newData = e.target.name.value;
    CharacterApiService.updateCharacter(
      this.state.characterId,
      'name',
      newData
    ).then((character) => {
      this.context.setCharacter(character);
      this.setState({ name: character.name, editingName: false });
    });
  };

  static contextType = CharacterContext;

  renderNavLinks = () => {
    let charErrors = [];
    let allScreens = [
      'race',
      'class',
      'background',
      'stats',
      'alignment',
      'character sheet'
    ];
    if (!this.context.character.race) {
      charErrors.push('race');
    }
    if (!this.context.character.class) {
      charErrors.push('class');
    }
    this.context.character.asi.forEach((asi) => {
      if (!asi.name && !charErrors.includes('stats')) {
        charErrors.push('stats');
      }
    });
    if (!charErrors.includes('stats')) {
      this.context.character.check_prof.forEach((prof) => {
        if (!prof.name && !charErrors.includes('stats')) {
          charErrors.push('stats');
        }
      });
    }

    if (!charErrors.includes('stats')) {
      Object.keys(this.context.character.abilities).forEach((ability) => {
        if (
          !this.context.character.abilities[ability].base &&
          !charErrors.includes('stats')
        ) {
          charErrors.push('stats');
        }
      });
    }

    if (!this.context.character.background) {
      charErrors.push('background');
    }

    const allNavLinks = allScreens.map((currscreen, index) => {
      let screen = currscreen.split(' ').join('');
      let charError = charErrors.includes(screen);

      return (
        <li
          ref={this.node}
          className={this.state.screen === screen && 'active'}
        >
          <Link to={screen} className="creator-nav-link" key={index}>
            {currscreen[0].toUpperCase() + currscreen.slice(1)}
            {charError ? (
              <sup className="alert">
                <i className="fas fa-exclamation"></i>
              </sup>
            ) : (
              <></>
            )}
          </Link>
        </li>
      );
    });

    return allNavLinks;
  };

  renderNextBack = () => {
    let allScreens = [
      'race',
      'class',
      'background',
      'stats',
      'alignment',
      'character sheet'
    ];
    const screenIndex = allScreens.findIndex(
      (screen) => this.state.screen === screen.split(' ').join('')
    );

    let nextScreen;
    let backScreen;
    if (screenIndex !== 0) {
      backScreen = allScreens[screenIndex - 1].split(' ').join('');
    }

    if (screenIndex !== allScreens.length - 1) {
      nextScreen = allScreens[screenIndex + 1].split(' ').join('');
    }

    const nextback = (
      <div className="nextback">
        {screenIndex !== 0 && (
          <Link to={`${backScreen}`}>
            <button className="nextback-button">{'Back'}</button>
          </Link>
        )}

        {screenIndex !== allScreens.length - 1 && (
          <Link tabIndex="1" to={`${nextScreen}`}>
            <button className="nextback-button">Next</button>
          </Link>
        )}
        <button ref={this.menu} className="hamburger" onClick={this.toggleNav}>
          <i className="fas fa-bars fa-lg"></i>
        </button>
        {this.state.showNav && (
          <ul className="nav-menu">
            <button
              className="close-nav"
              onClick={this.toggleNav}
              alt="close menu"
            >
              <i className="fas fa-times"></i>
            </button>
            {this.renderNavLinks()}
          </ul>
        )}
      </div>
    );
    return nextback;
  };

  render() {
    if (this.context.character) {
      return (
        <div className="spacer">
          <div className="character-creator-nav">
            <div className="creator-nav-content">
              <header className="character-creator-header">
                {!this.state.editingName && (
                  <h2>
                    {this.context.character.name}
                    <sup onClick={() => this.handleNameClick()}>
                      <i className="far fa-edit"></i>
                    </sup>
                  </h2>
                )}
                {this.state.editingName && (
                  <form onSubmit={(e) => this.handleNameSubmit(e)}>
                    <input
                      type="text"
                      id="name"
                      onChange={(e) => this.handleNameChange(e)}
                      value={this.state.name}
                      className="name-edit"
                    />
                  </form>
                )}
              </header>
              <nav className="nav-area">{this.renderNextBack()}</nav>
            </div>
          </div>
        </div>
      );
    } else {
      return <></>;
    }
  }
}

export default CreatorNav;
