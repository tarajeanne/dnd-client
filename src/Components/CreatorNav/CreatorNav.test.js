import React from 'react';
import CreatorNav from './CreatorNav';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Nav', () => {
  it('renders as expected', () => {
    const context = {
      character: {
        name: 'Sabrill',
        ac: 0,
        race: '',
        class: '',
        physical_desc: '',
        other_desc: '',
        background: '',
        alignment: '',
        size: '',
        speed: 0,
        languages: [
          {
            name: 'Language: Common',
            variable: false,
            depends_on: null
          }
        ],
        other_prof: [],
        asi: [],
        tool_prof: [],
        weap_prof: [],
        save_prof: [],
        check_prof: [],
        armor_prof: [],
        skills_and_features: [],
        prof_bonus: 2,
        hitDice: '',
        hp: 0,
        abilities: {
          constitution: {
            base: 0,
            total: 0,
            mod: 0
          },
          strength: {
            base: 0,
            total: 0,
            mod: 0
          },
          dexterity: {
            base: 0,
            total: 0,
            mod: 0
          },
          charisma: {
            base: 0,
            total: 0,
            mod: 0
          },
          intelligence: {
            base: 0,
            total: 0,
            mod: 0
          },
          wisdom: {
            base: 0,
            total: 0,
            mod: 0
          }
        },
        ability_checks: {
          acrobatics: {
            mod: 0,
            prof: 0,
            total: 0
          },
          'animal handling': {
            mod: 0,
            prof: 0,
            total: 0
          },
          arcana: {
            mod: 0,
            prof: 0,
            total: 0
          },
          athletics: {
            mod: 0,
            prof: 0,
            total: 0
          },
          deception: {
            mod: 0,
            prof: 0,
            total: 0
          },
          history: {
            mod: 0,
            prof: 0,
            total: 0
          },
          insight: {
            mod: 0,
            prof: 0,
            total: 0
          },
          intimidation: {
            mod: 0,
            prof: 0,
            total: 0
          },
          investigation: {
            mod: 0,
            prof: 0,
            total: 0
          },
          medicine: {
            mod: 0,
            prof: 0,
            total: 0
          },
          nature: {
            mod: 0,
            prof: 0,
            total: 0
          },
          perception: {
            mod: 0,
            prof: 0,
            total: 0
          },
          performance: {
            mod: 0,
            prof: 0,
            total: 0
          },
          persuasion: {
            mod: 0,
            prof: 0,
            total: 0
          },
          religion: {
            mod: 0,
            prof: 0,
            total: 0
          },
          'sleight of hand': {
            mod: 0,
            prof: 0,
            total: 0
          },
          stealth: {
            mod: 0,
            prof: 0,
            total: 0
          },
          survival: {
            mod: 0,
            prof: 0,
            total: 0
          }
        },
        weapons: [],
        armor: [],
        equipment: [],
        spells_known: [],
        spell_slots: [],
        spells_memorized: []
      }
    };
    const match = {
      path: 'http://localhost:3000/character/cjzc2ny7q0000fep28oy727gq/stats'
    };
    const wrapper = shallow(<CreatorNav match={match} />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
