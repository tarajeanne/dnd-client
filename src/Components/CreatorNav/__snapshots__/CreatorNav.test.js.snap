// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders as expected 1`] = `<Fragment />`;
